From: =?UTF-8?q?Marco=20Trevisan=20=28Trevi=C3=B1o=29?= <mail@3v1n0.net>
Date: Tue, 18 Apr 2023 22:07:37 +0200
Subject: display: Set compositor selection earlier on XWayland

When the X11 display is actually XWayland there's no point to delay the
compositor selection, given that mutter itself is the compositor and
doing this may cause the first X11 client that starts not to receive the
right information (and in some cases misbehave).

Since some toolkits are not handling the compositor selection changes
properly at later times, let's make their life easier by just
initializing the selection as early as the other X11 properties, given
that in this case there's nothing to replace.

This patch was grabbed from:
https://gitlab.gnome.org/3v1n0/mutter/-/commit/843b9c8d

Bug: https://gitlab.gnome.org/GNOME/mutter/-/issues/2472
Bug-Ubuntu: https://launchpad.net/bugs/1987976, https://launchpad.net/bugs/2016918
---
 src/core/display.c         |  1 -
 src/x11/meta-x11-display.c | 10 ++++++++++
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/src/core/display.c b/src/core/display.c
index 0ee95c0f0..eca1c0fcc 100644
--- a/src/core/display.c
+++ b/src/core/display.c
@@ -762,7 +762,6 @@ meta_display_init_x11_finish (MetaDisplay   *display,
   if (!display->display_opening)
     {
       g_signal_emit (display, display_signals[X11_DISPLAY_OPENED], 0);
-      meta_x11_display_set_cm_selection (x11_display);
       meta_display_manage_all_xwindows (display);
       meta_compositor_redirect_x11_windows (display->compositor);
     }
diff --git a/src/x11/meta-x11-display.c b/src/x11/meta-x11-display.c
index 3efd81acd..892f03230 100644
--- a/src/x11/meta-x11-display.c
+++ b/src/x11/meta-x11-display.c
@@ -736,6 +736,11 @@ take_manager_selection (MetaX11Display *x11_display,
     {
       XEvent event;
 
+#ifdef HAVE_XWAYLAND
+      if (meta_is_wayland_compositor ())
+        g_return_val_if_reached (new_owner);
+#endif
+
       /* We sort of block infinitely here which is probably lame. */
 
       meta_verbose ("Waiting for old window manager to exit");
@@ -1430,6 +1435,11 @@ meta_x11_display_new (MetaDisplay  *display,
   x11_display->wm_sn_atom = wm_sn_atom;
   x11_display->wm_sn_timestamp = timestamp;
 
+#ifdef HAVE_XWAYLAND
+  if (meta_is_wayland_compositor ())
+    meta_x11_display_set_cm_selection (x11_display);
+#endif
+
   init_event_masks (x11_display);
 
   meta_x11_display_init_frames_client (x11_display);
